### PR TITLE
Cake build system

### LINKED ISSUE #195: Split nuget packages
- Perspex.Core
- Perspex.ReactiveUI
- Perspex.Diagnostics
- Perspex.HtmlRenderer

### PR DESCRIPTION
**Pull request by wieslawsoltes**
Adds support for `Cake` build system: http://cakebuild.net/

Changes and new functionality:
- Added `build.ps1`  (Windows) and `build.sh` (Linux/OSX) bootstrapper scripts for Cake.
- Added `build.cake` Cake build script.
- Can build on CI server and locally (Linux/OSX/Windows).
- Splited Avalonia to separate NuGet packages, fixes #195 (the main NuGet package name remain the same). More information is included in `NuGet.md` file in PR.
- Build script is written in C# DSL.
- Automatic creation on `NuGet` packages integrated in `build.cake` script (no .nuspec necessary).
- Automatic publishing of `NuGet` packages to `MyGet` on `AppVeyor` (works same as before)
- Automatic publishing of `NuGet` packages to `NuGet` on `AppVeyor`. To make a release create a new `Release`  or `Tag` on `GitHub` with semantic version number. (TODO: Set `NUGET_API_KEY` in `appveyor.yml`).
- Automatic semantic versioning based on build configuration and place of build.

Sample command-line:
- Windows PowerShell: `.\build.ps1 -Target "Build" -Platform "Any CPU" -Configuration "Release"`
- Linux/OSX: `./build.sh --target "Travis" --platform "Mono" --configuration "Release"`

other possible targets:
- `Clean`
- `Build`
- `Run-Unit-Tests`
- `Create-NuGet-Packages or Package`
- `Publish-MyGet`
- `Publish-NuGet`
- `AppVeyor` (for CI use)
- `Travis` (for CI use)

### CODE REVIEW

**Comment by grokys** [comment]
This is an amazing PR @wieslawsoltes ! It's going to take me a little while to work through it, and I'm on vacation next week so it will probably be the week after before it gets fully reviewed and merged. Sorry about that!

File: appveyor.yml [hunk 1]
```diff
@@ -1,31 +1,25 @@
...
+- Any CPU
+configuration:
+- Release
 environment:
-  myget_key:
+  NUGET_API_KEY:
+    secure: ADD_SECURE_NUGET_API_KEY_HERE
+  NUGET_API_URL: https://www.nuget.org/api/v2/package
+  MYGET_API_KEY:
     secure: XOgD5bJUKNOS2kDDgb+affS4pDcslxALh+xvvnr1Koy0PjXlhILsBdNhxRe0KcNm
```
Comment (grokys) [inline appveyor.yml]: Oh! I didn't realise we had this in the repo. That's bad isn't it? Means someone could publish malicious packages to our MyGet feed.

Comment (wieslawsoltes) [inline appveyor.yml:11]: Not bad.

https://www.appveyor.com/docs/build-configuration/#secure-variables

"AppVeyor generates a unique encryption key for every account. To encrypt variable values go to Account -> Encrypt data tool."

The api keys are encrypted and tied to appveyor account.

We can remove those and set them through appveyor setting panel in neccessery.

Comment (grokys) [inline appveyor.yml]: Ah ok, so it's not the actual MyGet key - phew! :)

**Comment by grokys** [comment]
> all you need to do is create a new Release on GitHub with semantic version number.

Do you mean add something to the "Releases" tab in GitHub? For that we'd already need a build.

Or do you mean "tag with a release in git"?

**Comment by wieslawsoltes** [comment]
Release creates a tag, so both ways should work.

**Comment by wieslawsoltes** [comment]
https://github.com/AvaloniaUI/Avalonia/pull/688/commits/1d2be074d6e6bcb1135083b00c8cba2811e35893#diff-4d3634747586b964df9f7712bc3ec669R56

Build scripts check for `Tag` and `Publish-NuGet` task checks if all conditions are met: https://github.com/AvaloniaUI/Avalonia/pull/688/commits/1d2be074d6e6bcb1135083b00c8cba2811e35893#diff-4d3634747586b964df9f7712bc3ec669R683

**Comment by grokys** [comment]
I had to remove Xamarin as it was causing problems with my day job ;), but now when I try to do a build using this I get an error:

```
D:\projects\Avalonia\src\Skia\Avalonia.Skia.Android\Avalonia.Skia.Android.csproj(98,3): error MSB4019: The imported pro
ject "C:\Program Files (x86)\MSBuild\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Confirm that the pa
th in the <Import> declaration is correct, and that the file exists on disk.
D:\projects\Avalonia\src\Skia\Avalonia.Skia.Android.TestApp\Avalonia.Skia.Android.TestApp.csproj(118,3): error MSB4019:
 The imported project "C:\Program Files (x86)\MSBuild\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Co
nfirm that the path in the <Import> declaration is correct, and that the file exists on disk.
D:\projects\Avalonia\src\Skia\Avalonia.Skia.iOS\Avalonia.Skia.iOS.csproj(95,3): error MSB4019: The imported project "C:
\Program Files (x86)\MSBuild\Xamarin\iOS\Xamarin.iOS.CSharp.targets" was not found. Confirm that the path in the <Impor
t> declaration is correct, and that the file exists on disk.
D:\projects\Avalonia\src\Android\Avalonia.Android\Avalonia.Android.csproj(142,3): error MSB4019: The imported project "
C:\Program Files (x86)\MSBuild\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Confirm that the path in
the <Import> declaration is correct, and that the file exists on disk.
D:\projects\Avalonia\src\Android\Avalonia.AndroidTestApplication\Avalonia.AndroidTestApplication.csproj(192,3): error M
SB4019: The imported project "C:\Program Files (x86)\MSBuild\Xamarin\Android\Xamarin.Android.CSharp.targets" was not fo
und. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
D:\projects\Avalonia\src\iOS\Avalonia.iOS\Avalonia.iOS.csproj(115,3): error MSB4019: The imported project "C:\Program F
iles (x86)\MSBuild\Xamarin\iOS\Xamarin.iOS.CSharp.targets" was not found. Confirm that the path in the <Import> declara
tion is correct, and that the file exists on disk.
```

I assume this is caused by not having Xamarin. I wonder - is there any way to test for the presence of Xamarin and disable those parts if necessary? Or some other way of building without it?

**Comment by wieslawsoltes** [comment]
@grokys Do not thinks there is elegant solution for this problem, only way I could find is to create new `Build Configuration`. See my PR https://github.com/AvaloniaUI/Avalonia/pull/693 and try building with `-Configuration "NoXamarin"`.

**Comment by wieslawsoltes** [comment]
And in build script we would have to check for `NoXamarin` configuration and exclude mobile part from task, like `NuGet` creation.

**Comment by grokys** [comment]
Yeah, it might not be worth it if it's not such an easy thing to do. Visual Studio handles it fine and that's where 90% of people will be building.

File: build.cake [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+    GetDirectories("./tests/**/bin/" + dirSuffix) + 
+    GetDirectories("./tests/**/obj/" + dirSuffix) + 
+    GetDirectories("./Samples/**/bin/" + dirSuffix) + 
+    GetDirectories("./Samples/**/obj/" + dirSuffix);
+
+///////////////////////////////////////////////////////////////////////////////
+// NUGET NUSPECS
+///////////////////////////////////////////////////////////////////////////////
+
+var SerilogVersion = "1.5.14";
```
Comment (kekekeks) [inline build.cake]: I think that should be extracted from packages.config from project (and trigger error if there is version mismatch in packages referenced from different projects)

**Pull request merged**